fdbserver/RestoreApplier.actor.cpp

```diff
@@ -216,47 +218,72 @@ ACTOR static Future<Void> getAndComputeStagingKeys(
     std::map<Key, std::map<Key, StagingKey>::iterator> incompleteStagingKeys, double delayTime, Database cx,
     UID applierID, int batchIndex) {
 	state Reference<ReadYourWritesTransaction> tr(new ReadYourWritesTransaction(cx));
-	state std::vector<Future<Optional<Value>>> fValues;
+	state std::vector<Future<Optional<Value>>> fValues(incompleteStagingKeys.size(), Never());
 	state int retries = 0;
+	// state UID randomID = deterministicRandom()->randomUniqueID();
 
 	wait(delay(delayTime + deterministicRandom()->random01() * delayTime));
 	TraceEvent("FastRestoreApplierGetAndComputeStagingKeysStart", applierID)
+	    //.detail("RandomUID", randomID)
 	    .detail("BatchIndex", batchIndex)
 	    .detail("GetKeys", incompleteStagingKeys.size())
 	    .detail("DelayTime", delayTime);
+	state std::set<int> keyNotFounds;
+
+	state int i = 0;
+	state bool hasError = false;
 	loop {
+		hasError = false;
 		try {
 			tr->reset();
 			tr->setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
 			tr->setOption(FDBTransactionOptions::LOCK_AWARE);
+			i = 0;
 			for (auto& key : incompleteStagingKeys) {
-				fValues.push_back(tr->get(key.first));
+				if (!fValues[i].isReady() || !keyNotFounds.count(i)) {
+					fValues[i] = tr->get(key.first);
+				}
+				++i;
+			}
+			for (i = 0; i < incompleteStagingKeys.size(); ++i) {
+				if (keyNotFounds.count(i)) {
+					continue;
+				}
+				wait(success(fValues[i])); // NOTE: This may be waiting for ever!
```

Comment:
This is doing sequential `wait` on all keys. I think it's possible to do a `waitForAll(fValues)`, where the `fValues` is set to `Void()` (instead of `Never()`) for non-existent keys.

Reply:
The challenge is how to identify the not-exist key efficiently and correctly.
Say we want to get key k1, k2, k3, where k2 does not exist. k2's `get()` throws an error before k3 is ready. When we use `waitForAll`, how do we know which key is not existent and which key should retry? 
I guess I can check the error state of the future. 

This seems to me a generic requirement: Get keys and return non-existent key as null. Do we have such a functionality in the code before I rebuild the wheel?

Reply:
I don't think current API allows an option to return non-existent key as null. In fact, the null/empty value is legal.

I think you can do:
```
ACTOR Future<Optional<Value>> getValue(tr, key, i, keyNotFounds) {
  try {
    Optional<Value> v = wait(tr->get(key));
    return v;
  } catch (Error e) {
    if (e == error_code_key_not_found) {
       keyNotFounds->insert(i);
       return Optional<Value>();
    } else {
       throw;
    }
  }
}

  ...
  for (i) {
    if (keyNotFounds.count(i) > 0) {
      fValues[i] = getValue(tr, key, i, &keyNotFounds);
    } else {
      fValues[i] = Void();
    }
  }

  waitForAll(fValues);

```